add channel last support for MaxUnpool2d

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58348 opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU
* #55864 enable BFloat16 mkldnn_convolution on both contiguous and channels last memory format
* #55584 add channels last (2d) support for mkldnn_convolution
* #51185 add channels last support for ConvTranspose2d
* #50573 add channels last support for PixelShuffle and PixelUnshuffle
* #50247 add channels last support for ChannelShuffle
* **#49984 add channel last support for MaxUnpool2d**
* #49821 add channels last for GroupNorm

Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

mingfeima committed:
add channel last support for MaxUnpool2d

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007051).

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007051).

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007051).

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007051).

VitalyFedyunin commented:
Benchmarks please (also please do not forget before - after ).

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima commented:
Update: add benchmark data
* avx512 CPU: Intel(R) Xeon(R) Gold 6248, dual sockets, 20 cores per socket, 2.50GHz
* avx2 CPU: Intel(R) Xeon(R) E5-2680 v3, dual sockets, 12 cores per socket, 2.50GHz

#### single socket inference (avx512)
```bash
before:
MaxUnPool2d(contiguous): input_size [1, 64, 112, 112] time: 0.369 ms
MaxUnPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.203 ms
MaxUnPool2d(contiguous): input_size [128, 64, 112, 112] time: 44.327 ms
MaxUnPool2d(channels_last): input_size [128, 64, 112, 112] time: 27.688 ms

after:
MaxUnPool2d(contiguous): input_size [1, 64, 112, 112] time: 0.367 ms
MaxUnPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.046 ms
MaxUnPool2d(contiguous): input_size [128, 64, 112, 112] time: 43.759 ms
MaxUnPool2d(channels_last): input_size [128, 64, 112, 112] time: 10.156 ms
```

#### single core inference (avx512)
```bash
before:
MaxUnPool2d(contiguous): input_size [1, 64, 112, 112] time: 5.480 ms
MaxUnPool2d(channels_last): input_size [1, 64, 112, 112] time: 1.417 ms
MaxUnPool2d(contiguous): input_size [128, 64, 112, 112] time: 696.883 ms
MaxUnPool2d(channels_last): input_size [128, 64, 112, 112] time: 258.977 ms

after:
MaxUnPool2d(contiguous): input_size [1, 64, 112, 112] time: 5.432 ms
MaxUnPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.532 ms
MaxUnPool2d(contiguous): input_size [128, 64, 112, 112] time: 699.355 ms
MaxUnPool2d(channels_last): input_size [128, 64, 112, 112] time: 79.025 ms
```

#### single socket inference (avx2)
```bash
before:
MaxUnPool2d(contiguous): input_size [1, 64, 112, 112] time: 1.420 ms
MaxUnPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.437 ms
MaxUnPool2d(contiguous): input_size [128, 64, 112, 112] time: 104.059 ms
MaxUnPool2d(channels_last): input_size [128, 64, 112, 112] time: 33.598 ms

after:
MaxUnPool2d(contiguous): input_size [1, 64, 112, 112] time: 1.403 ms
MaxUnPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.143 ms
MaxUnPool2d(contiguous): input_size [128, 64, 112, 112] time: 115.405 ms
MaxUnPool2d(channels_last): input_size [128, 64, 112, 112] time: 13.674 ms
```

#### single core inference (avx2)
```bash
before:
MaxUnPool2d(contiguous): input_size [1, 64, 112, 112] time: 7.880 ms
MaxUnPool2d(channels_last): input_size [1, 64, 112, 112] time: 1.480 ms
MaxUnPool2d(contiguous): input_size [128, 64, 112, 112] time: 895.136 ms
MaxUnPool2d(channels_last): input_size [128, 64, 112, 112] time: 317.684 ms

after:
MaxUnPool2d(contiguous): input_size [1, 64, 112, 112] time: 7.788 ms
MaxUnPool2d(channels_last): input_size [1, 64, 112, 112] time: 0.771 ms
MaxUnPool2d(contiguous): input_size [128, 64, 112, 112] time: 900.540 ms
MaxUnPool2d(channels_last): input_size [128, 64, 112, 112] time: 118.523 ms
```

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

mingfeima committed:
Update on "add channel last support for MaxUnpool2d"


Differential Revision: [D26007051](https://our.internmc.facebook.com/intern/diff/D26007051)

[ghstack-poisoned]

VitalyFedyunin commented on test/test_nn.py:
```diff
@@ -6097,6 +6097,37 @@ def test_MaxUnpool2d_output_size(self):
 
+    def test_max_unpool2d_nhwc_cpu(self):
+        input = torch.randn(2, 10, 9, 9).float().cpu()
+        input = input.contiguous(memory_format=torch.channels_last)
```
nit: please add tests for randomly permuted inputs

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D26007051).
